srv/modules/runners/openstack.py

```diff
@@ -0,0 +1,212 @@
+    """
+    local = salt.client.LocalClient()
+
+    __opts__ = salt.config.client_config('/etc/salt/master')
+    __grains__ = salt.loader.grains(__opts__)
+    __opts__['grains'] = __grains__
+    __utils__ = salt.loader.utils(__opts__)
+    __salt__ = salt.loader.minion_mods(__opts__, utils=__utils__)
+    master_minion = __salt__['master.minion']()
+
+    if "prefix" in kwargs:
+        state_res = local.cmd(master_minion, 'state.apply', ['ceph.openstack',
+                              'pillar={"openstack_prefix": "' + kwargs['prefix'] + '"}'])
+    else:
+        state_res = local.cmd(master_minion, 'state.apply', ['ceph.openstack'])
+
+    # If state.apply failed for any reason, this will return whatever
+    # state(s) failed to apply
+    failed = []
+    for _, states in state_res.items():
+        if isinstance(states, dict):
+            for _, state in states.items():
+                if 'result' not in state or not state['result']:
+                    failed.append(state)
+        else:
+            # This could happen if the SLS being applied somehow doesn't exist,
+            # e.g. "No matching sls found for 'ceph.openstack' in env 'base'".
+            # Realistically this should never /actually/ happen.
+            failed.append(states)
+    if failed:
+        return {'ERROR': failed}
+
+    runner = salt.runner.RunnerClient(__opts__)
+
+    # Set up prefix for subsequent string concatenation to match what's done
+    # in the SLS files for keyring and pool names.
+    prefix = ""
+    if "prefix" in kwargs:
+        prefix = kwargs['prefix'] + "-"
+
+    conf = configparser.RawConfigParser()
+    with open("/srv/salt/ceph/configuration/cache/ceph.conf") as lines:
+        conf.read_string('\n'.join(line.strip() for line in lines))
+
+    radosgw_url = ""
+
+    rgw_urls = []
+    rgw_configurations = runner.cmd('select.from',
+                                    ['pillar=rgw_configurations', 'role=rgw', 'attr=host'],
+                                    print_event=False)
+    for rgw in rgw_configurations:
+        section = "client." + rgw[0] + "." + rgw[1]
+        if not conf.has_section(section):
+            continue
+        if conf.has_option(section, "rgw frontends") and conf.has_option(section, "rgw dns name"):
+            https = re.match(r'.*port=(\d+)s.*', conf.get(section, "rgw frontends"))
+            http = re.match(r'.*port=(\d+).*', conf.get(section, "rgw frontends"))
+            if not http and not https:
+                continue
+            if http:
+                url = "http://" + conf.get(section, "rgw dns name") + ":" + http.group(1)
+            if https:
+                url = "https://" + conf.get(section, "rgw dns name") + ":" + https.group(1)
+            url += "/swift/v1"
+            rgw_urls.append(url)
+    if len(rgw_urls) == 1:
+        radosgw_url = rgw_urls[0]
+    elif len(rgw_urls) > 1:
+        # TODO: This is...  Not ideal.  But I don't have any better ideas right now.
```

Comment:
You mean that you don't have a good way to distinct which rgw_url to return?

Reply:
Correct. That the string comes back as "FIXME: http://rgw1/ http://rgw2 ..." at least is easy for a human reading to know they need to make a choice, and is easy for an automation tool to match and fail on. If that's OK (or if we don't come up with anything better), then I'll remove the "TODO".

Hrm.  I wonder if I should maybe change it to just return a list of all RGW endpoints, rather than a weird string.

Reply:
I think making a list of all of the RGW endpoints would be better here.   The user can edit them if needed, or the integration in SOC will just pick the first in the list.